geonetwork/core-geonetwork
# Metadata workflow - display error when creating a working copy without download privilege

Editing an approved metadata with a user who has the `Edit` privilege on the metadata, but does not have the `Download` privilege leaves the metadata in an inconsistent state.

This change prevents this, displaying an error to the user when tries to edit the approved metadata.

Test case:

1) Login as admin and enable the metadata workflow.
2) Create an editor user in the `Sample` group.
3) Create an iso19139 metadata with the default template and set the publication date.
4) Approve the metadata.
5) In the privileges dialog enable the `Edit` privilege, but not the `Download` privilege.


6) Login as the editor user created in step 2 and edit the metadata changing the publication date. Click `Save and close` the metadata editor --> The metadata detail page is inconsistent, showing the publication date swapped between the approved and working copy pages.

In the log file when editing the metadata, an exception is logged, but it is silently swallowed, causing the problem:

```
2022-08-18 14:47:40,176 ERROR [geonetwork.datamanager] - Editing instance creation failed
org.eclipse.jetty.io.RuntimeIOException: java.lang.SecurityException: Current user can't download resources for metadata '73066f1c-be70-47cc-a6a8-80cc4d3e5a01' and as such can't access the requested resource.
	at org.fao.geonet.kernel.datamanager.draft.DraftMetadataUtils.cloneFiles(DraftMetadataUtils.java:602)
```

With this change the working copy is not created and the following error message is displayed to the user.

![error-message](https://user-images.githubusercontent.com/1695003/185399871-80394e26-c6ec-4f1b-a514-82cd209ee17c.png)

---

To backport the PR to `3.12.x` the javascript change is not required, seems that `3.12.x` version returns xml for the errors, but `4.x` version, at least in some cases, seems to return json.
